### ISSUE TITLE
Proposal: Drop Python 2 support in v0.20.0

### ISSUE THREAD

**Issue by tgaddair** [horovod/horovod#1929]
At this point, most frameworks in our ecosystem have dropped Python 2 support, so it makes sense for us to consider doing the same.  

Already, a number of our more recent features (elastic mode, synchronized batch normalization) are Python 3 only.

We still officially maintain support for Python 2 for all legacy code paths, but there are a number of reasons to consider dropping this:

1. Running Python 2 unit tests cost money to run in the cloud, and since we have nearly 1:1 Python 2 to Python 3 test coverage, this nearly doubles our cloud bill for CPU tests.
2. As new versions of Python continue to be released, it becomes a challenge to just maintain mutual compatibility between Python 3.6 (our earliest supported version of Python 3) and newer versions (python 3.8, which recently changes the default multiprocessing run mode, see #1891). 
3. Maintaining feature parity becomes a challenge, as a number of basic things supported natively in Python 3 require elaborate workarounds for Python 2 (for example, multiprocessing barrier).

Additionally, users who need to use Python 2 can continue to use older versions of Horovod.  As TensorFlow and PyTorch should not be releasing new Python 2 versions of their libraries, this shouldn't be an issue for users who are stuck using Python 2.

What does dropping Python 2 support mean?

1. No more Python 2 tests in Buildkite.
2. No more Python 2 Docker images or PyPi packages.
3. No more system version checks to perform different logic for Python 2 vs Python 3.
4. No more dependency on `six`, replace all usages with native Python 3 equivalents.
5. No more doc references to Python 2, or unit test exclusions based on Python 2.

TSC members, please review and provide your feedback:

@alsrgv @nvcastet @romerojosh @EnricoMi @karakusc @apeforest

**Comment by alsrgv** [comment]
+1 for removing Python 2 support for future releases.

**Comment by DEKHTIARJonathan** [comment]
+1 I also support this move ;)

**Comment by EnricoMi** [comment]
+1 for finally removing Python 2 ... and no more `from __future__ import ...` workarounds

**Comment by romerojosh** [comment]
+1 I also support this!

**Comment by karakusc** [comment]
+1 agree with the reasoning

**Comment by nvcastet** [comment]
+1 yes that makes complete sense.
